share/normalize/mods.py

```diff
@@ -0,0 +1,431 @@
+        typeOfResource = tools.Try(ctx['mods:typeOfResource'])
+
+    def check_status(self, status):
+        if status == 'deleted':
+            return True
+        return False
+
+    # TODO
+    # should check current models if no type map is given
+    # either use registries attribute of abstractcreativework or check if model exists
+    def get_schema(self, types):
+        if not types or not self.type_map:
+            return self.default_type
+        if isinstance(types, str):
+            types = [types]
+        for t in types:
+            if isinstance(t, dict):
+                t = t['#text']
+            t = t.lower()
+            if t in self.type_map:
+                return self.type_map[t]
+        return self.default_type
+
+    def tokenize(self, data):
+        if isinstance(data, str):
+            data = [data]
+        tokens = []
+        for item in data:
+            tokens.extend([x.strip() for x in re.split(r'(?: - )|\.|,', item) if x])
+        return tokens
+
+
+class MODSNormalizer(Normalizer):
+
+    @property
+    def root_parser(self):
+        class RootParser(MODSCreativeWork):
+            default_type = self.config.emitted_type.lower()
+            type_map = {
+                **{r.lower(): r for r in self.allowed_roots},
+                **{t.lower(): v for t, v in self.config.type_map.items()}
+            }
+
+        return RootParser
+
+    def do_normalize(self, data):
+        if self.config.approved_sets is not None:
+            specs = set(x.replace('publication:', '') for x in etree.fromstring(data).xpath(
+                'ns0:header/ns0:setSpec/node()',
+                namespaces={'ns0': 'http://www.openarchives.org/OAI/2.0/'}
+            ))
+            if not (specs & set(self.config.approved_sets)):
+                logger.warning('Series %s not found in approved_sets for %s', specs, self.config.label)
+                return None
+
+        return super().do_normalize(data)
+
+    def unwrap_data(self, data):
+        unwrapped_data = xmltodict.parse(data, process_namespaces=True, namespaces=self.namespaces)
+
+        flattened_data = {}
+        flattened_data['header'] = unwrapped_data['record']['header']
+        top_level_fields = [
+            'mods:abstract',
+            'mods:accessCondition',
+            'mods:classification',
+            'mods:extension',
+            'mods:genre',
+            'mods:identifier',
+            'mods:language',
+            'mods:location',
+            'mods:name',
+            'mods:note',
+            'mods:originInfo',
+            'mods:part',
+            'mods:physicalDescription',
+            'mods:recordInfo',
+            'mods:relatedItem',
+            'mods:subject',
+            'mods:tableOfContents',
+            'mods:targetAudience',
+            'mods:titleInfo',
+            'mods:typeOfResource'
+        ]
+
+        for field in top_level_fields:
```

Comment:
Could this just be `flattened_data = {**unwrapped_data['record']['metadata']['mods:mods']}`